File: java/hypertable-common/src/main/java/org/hypertable/FsBroker/Lib/StatusManager.java
/*
 * Copyright (C) 2007-2015 Hypertable, Inc.
 *
 * This file is part of Hypertable.
 *
 * Hypertable is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 3
 * of the License, or any later version.
 *
 * Hypertable is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301, USA.
 */


package org.hypertable.FsBroker.Lib;

import org.hypertable.Common.Status;

import java.io.IOException;
import java.net.InetSocketAddress;
import java.util.HashMap;
import java.util.Iterator;
import java.util.Map;

/**
 * This class acts as a synchronized HashMap that maps open file descriptors to
 * OpenFileData objects
 */
public class StatusManager {

  /** Constructor.
   */
  public StatusManager() {
    mStatus = new Status();
    mCurrentStatus = mStatus.get();
  }

  public synchronized void setReadException(Exception e) {
    Status.Code code = Status.Code.CRITICAL;
    if (mCurrentStatus == Status.Code.OK || mCurrentStatus != Status.Code.CRITICAL) {
      mCurrentStatus = code;
      mStatus.set(code, e.toString());
    }
    mLastReadError = System.currentTimeMillis();
  }

  public synchronized void setWriteException(Exception e) {
    Status.Code code = Status.Code.CRITICAL;
    if (mCurrentStatus == Status.Code.OK || mCurrentStatus != Status.Code.CRITICAL) {
      mCurrentStatus = code;
      mStatus.set(code, e.toString());
    }
    mLastWriteError = System.currentTimeMillis();
  }

  public void setReadStatus(Status.Code code, String text) {
    if (code == Status.Code.OK) {
      clearStatus();
      return;
    }
    synchronized (this) {
      if (mCurrentStatus == Status.Code.OK ||
          (mCurrentStatus != Status.Code.CRITICAL && code == Status.Code.CRITICAL)) {
        mCurrentStatus = code;
        mStatus.set(code, text);
      }
      mLastReadError = System.currentTimeMillis();
    }
  }

  public void setWriteStatus(Status.Code code, String text) {
    if (code == Status.Code.OK) {
      clearStatus();
      return;
    }
    synchronized (this) {
      if (mCurrentStatus == Status.Code.OK ||
          (mCurrentStatus != Status.Code.CRITICAL && code == Status.Code.CRITICAL)) {
        mCurrentStatus = code;
        mStatus.set(code, text);
      }
      mLastWriteError = System.currentTimeMillis();
    }
  }

  public void clearStatus() {
    if (mCurrentStatus == Status.Code.OK)
      return;
    long now = System.currentTimeMillis();
    synchronized (this) {
      if (mCurrentStatus == Status.Code.WARNING ||
          (now - mLastReadError > CLEAR_CHANGE_INTERVAL &&
           now - mLastWriteError > CLEAR_CHANGE_INTERVAL)) {
        mStatus.set(Status.Code.OK, null);
        mCurrentStatus = Status.Code.OK;
      }
    }
  }

  public Status get() { return mStatus; }

  private final long CLEAR_CHANGE_INTERVAL = 60000;
  private Status mStatus;
  private long mLastReadError;
  private long mLastWriteError;
  private Status.Code mCurrentStatus;
}

File: src/cc/FsBroker/Lib/StatusManager.cc
/* -*- c++ -*-
 * Copyright (C) 2007-2015 Hypertable, Inc.
 *
 * This file is part of Hypertable.
 *
 * Hypertable is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 3
 * of the License, or any later version.
 *
 * Hypertable is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301, USA.
 */

/// @file
/// Definitions for StatusManager.
/// This file contains definitions for StatusManager, a class for managing the
/// status of a file system broker.

#include <Common/Compat.h>

#include "StatusManager.h"

#include <Common/String.h>

#include <cstring>

using namespace Hypertable;
using namespace Hypertable::FsBroker::Lib;
using namespace std;

namespace {
  // Minimum elapsed time before transitioning back to OK status
  const chrono::steady_clock::duration CLEAR_CHANGE_INTERVAL = chrono::milliseconds(60000);
}

StatusManager::StatusManager() {
  m_current_status = m_status.get();
}

void StatusManager::set_read_status(Status::Code code, const std::string &text) {
  if (code == Status::Code::OK)
    clear_status();
  else {
    set_status(code, text);
    m_last_read_error = chrono::steady_clock::now();
  }
}

void StatusManager::set_write_status(Status::Code code, const std::string &text) {
  if (code == Status::Code::OK)
    clear_status();
  else {
    set_status(code, text);
    m_last_write_error = chrono::steady_clock::now();
  }
}

void StatusManager::set_read_error(int error) {
  if (error == 0)
    clear_status();
  else {
    set_error(error);
    m_last_read_error = chrono::steady_clock::now();
  }
}

void StatusManager::set_write_error(int error) {
  if (error == 0)
    clear_status();
  else {
    set_error(error);
    m_last_write_error = chrono::steady_clock::now();
  }
}

void StatusManager::clear_status() {
  if (m_current_status == Status::Code::OK)
    return;
  auto now = chrono::steady_clock::now();
  lock_guard<mutex> lock(m_mutex);
  if (now - m_last_read_error > CLEAR_CHANGE_INTERVAL &&
      now - m_last_write_error > CLEAR_CHANGE_INTERVAL) {
    m_status.set(Status::Code::OK, "");
    m_current_status = Status::Code::OK;
  }
}

void StatusManager::set_status(Status::Code code, const std::string &text) {
  lock_guard<mutex> lock(m_mutex);
  m_current_status = code;
  m_status.set(m_current_status, text);
}

void StatusManager::set_error(int error) {
  char errtext[128];
  errtext[0] = 0;
  strerror_r(errno, errtext, 128);
  lock_guard<mutex> lock(m_mutex);
  m_current_status = Status::Code::CRITICAL;
  m_status.set(m_current_status, errtext);
}
